fs/io_zenfs.cc

```diff
@@ -515,6 +515,100 @@ IOStatus ZoneFile::Append(void* data, int data_size) {
   return IOStatus::OK();
 }
 
+IOStatus ZoneFile::RecoverSparseExtents(uint64_t start, uint64_t end,
+                                        Zone* zone) {
+  /* Sparse writes, we need to recover each individual segment */
+  IOStatus s;
+  uint32_t block_sz = GetBlockSize();
+  int f = zbd_->GetReadFD();
+  uint64_t next_extent_start = start;
+  char* buffer;
+  int recovered_segments = 0;
+  int ret;
+
+  ret = posix_memalign((void**)&buffer, sysconf(_SC_PAGESIZE), block_sz);
+  if (ret) {
+    return IOStatus::IOError("Out of memory while recovering");
+  }
+
+  while (next_extent_start < end) {
+    uint64_t extent_length;
+
+    ret = pread(f, (void*)buffer, block_sz, next_extent_start);
+    if (ret != (int)block_sz) {
+      s = IOStatus::IOError("Unexpected read error while recovering");
+      break;
+    }
+
+    extent_length = DecodeFixed64(buffer);
+    if (extent_length == 0) {
+      s = IOStatus::IOError("Unexexpeted extent length while recovering");
+      break;
+    }
+    recovered_segments++;
+
+    zone->used_capacity_ += extent_length;
+    extents_.push_back(new ZoneExtent(next_extent_start + SPARSE_HEADER_SIZE,
+                                      extent_length, zone));
+
+    uint64_t extent_blocks = 1 + extent_length / block_sz;
+    next_extent_start += extent_blocks * block_sz;
```

Comment:
I think it would be easier to understand like that:
```suggestion
    next_extent_start += extent_length + block_sz;
```

Reply:
Hmm , the suggested change is not equivalent in behavior. The next extent starts on the next block boundary.

Reply:
Maybe I am missing something here, but the suggestion should be equivalent (given that `extent_length` is a multiple of `block_sz`):
`(1 + extent_length / block_sz) * block_sz == block_sz * 1 + block_sz * extent_length / block_sz == block_sz + extent_length`

Reply:
Oh wait, I think that I got your point. `extent_length` is not block aligned (thus not a multiple of `block_sz`) and therefore you are right.